clients/src/test/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManagerTest.java

```diff
@@ -0,0 +1,675 @@
+        expectSuccessfulRequest(Collections.singletonMap(TEST_PARTITION_1, LEADER_1));
+        requestManager.onUpdate(new ClusterResource(""));
+        assertEquals(1, requestManager.requestsToSend());
+
+        Map<TopicPartition, OffsetAndTimestamp> expectedOffsets = Collections.singletonMap(
+                TEST_PARTITION_1, new OffsetAndTimestamp(5L, 1L));
+        verifySuccessfulPollAndResponseReceived(fetchOffsetsFuture, expectedOffsets);
+    }
+
+    @ParameterizedTest
+    @MethodSource("retriableErrors")
+    public void testRequestFailsWithRetriableError_RetrySucceeds(Errors error) throws ExecutionException, InterruptedException {
+        Map<TopicPartition, Long> timestampsToSearch = Collections.singletonMap(TEST_PARTITION_1,
+                ListOffsetsRequest.EARLIEST_TIMESTAMP);
+
+        // List offsets request successfully built
+        expectSuccessfulRequest(Collections.singletonMap(TEST_PARTITION_1, LEADER_1));
+        CompletableFuture<Map<TopicPartition, OffsetAndTimestamp>> fetchOffsetsFuture = requestManager.fetchOffsets(
+                timestampsToSearch,
+                false);
+        assertEquals(1, requestManager.requestsToSend());
+        assertEquals(0, requestManager.requestsToRetry());
+
+        // Request successfully sent to single broker
+        NetworkClientDelegate.PollResult res = requestManager.poll(time.milliseconds());
+        verifySuccessfulPollAwaitingResponse(res);
+        assertFalse(fetchOffsetsFuture.isDone());
+
+        // Response received with error
+        NetworkClientDelegate.UnsentRequest unsentRequest = res.unsentRequests.get(0);
+        ClientResponse clientResponse = buildClientResponseWithErrors(
+                unsentRequest,
+                Collections.singletonMap(TEST_PARTITION_1, error));
+        clientResponse.onComplete();
+        assertFalse(fetchOffsetsFuture.isDone());
+        assertEquals(1, requestManager.requestsToRetry());
+        assertEquals(0, requestManager.requestsToSend());
+
+        // Cluster metadata update. Failed requests should be retried and succeed
+        expectSuccessfulRequest(Collections.singletonMap(TEST_PARTITION_1, LEADER_1));
+        requestManager.onUpdate(new ClusterResource(""));
+        assertEquals(1, requestManager.requestsToSend());
+
+        Map<TopicPartition, OffsetAndTimestamp> expectedOffsets = Collections.singletonMap(TEST_PARTITION_1, new OffsetAndTimestamp(5L, 1L));
+        verifySuccessfulPollAndResponseReceived(fetchOffsetsFuture, expectedOffsets);
+    }
+
+    @Test
+    public void testRequestNotSupportedErrorReturnsNullOffset() throws ExecutionException,
+            InterruptedException {
+        testResponseWithErrorAndUnknownOffsets(Errors.UNSUPPORTED_FOR_MESSAGE_FORMAT);
+    }
+
+    @Test
+    public void testRequestWithUnknownOffsetInResponseReturnsNullOffset() throws ExecutionException,
+            InterruptedException {
+        testResponseWithErrorAndUnknownOffsets(Errors.NONE);
```

Comment:
This works, but is a bit unintuitive since Errors.NONE doesn't represent an error.

Reply:
Agree. Given that the logic is the same I will opt for renaming it to "error code" instead of "error", to better group Errors.NONE and Errors.UNSUPPORTED_FOR_MESSAGE_FORMAT. So including `testResponseWithErrorCodeAndUnknownOffsets`